### DISCUSSION TITLE
Clarification on numeric literals

### DISCUSSION THREAD

**Discussion by benjreinhart** [kdl-org/kdl#171] [Q&A] [answered]
Hello... I am looking for some clarification around a couple things related to numbers.

(Apologize in advance if this is obvious to everyone, I am new to implementing parsers and interpreting language specs).

1. For numbers that specify a radix, the radix MUST be followed by a respective digit, correct? For example, an underscore cannot immediately follow `0b`, `0o`, and/or `0x`, right (e.g., `0b_01` is illegal)?
1. If there was a trailing underscore, the lexer would NOT include that in the number literal, but instead probably treat that as the start of an identifier (which I think is then illegal from the parser's POV), e.g., `0b01_`?
1. Only a single underscore is valid in between digits, so `123___456_789` would be illegal?
1. In decimal numbers, can an underscore appear after an exponent denoted with `e` or `E`?

If folks think it would be helpful to make any of the above worded more explicitly in the spec, I'd be happy to take a stab at that.

🙏🏻

**Comment by zkat** [comment, marked as answer]
Spec ref:
```
decimal := integer ('.' [0-9] [0-9_]*)? exponent?
exponent := ('e' | 'E') integer
integer := sign? [0-9] [0-9_]*
sign := '+' | '-'

hex := sign? '0x' hex-digit (hex-digit | '_')*
octal := sign? '0o' [0-7] [0-7_]*
binary := sign? '0b' ('0' | '1') ('0' | '1' | '_')*

```

Translation:

1. Correct, that is illegal. (`'0x' hex-digit (hex-digit | '_')*`)
2. Trailing underscores are a valid part of numbers
3. You can have multiple underscores.
4. Yes, after the first post-E digit (see definition for `exponent` above)
